numpy/core/src/npysort/selection.c.src

```diff
@@ -406,6 +448,89 @@ NPY_NO_EXPORT int
 }
 
 
+/*
+ * rivest-floyd selection algorithm
+ */
+#define C1 600
+#define C2 0.5
+#define C3 0.5
+
+/* helper function */
+void
+@name@rf_recursive_@suff@(@type@ *v,
+#if @arg@
+                          npy_intp* tosort,
+#endif
+                          npy_intp left,
+                          npy_intp right,
+                          npy_intp kth)
+{
+    npy_intp j;
+    while (right > left) {
+        if (right - left > C1) {
+            npy_float N = (npy_float)(right - left + 1);
+            npy_float II = (npy_float)(kth - left + 1);
+            npy_float Z = npy_log(N);
+            npy_float S = C2 * npy_exp(2*Z/3.0);
+            npy_float SD = C3 * npy_sqrt(Z*S*(N - S)/N) * npy_copysign(1, II - N/2);
+            npy_intp newleft = kth - (npy_intp)(II*S/N + SD);
+            if (newleft < left) newleft = left;
+            npy_intp newright = kth + (npy_intp)((N - II)*S/N + SD);
+            if (newright > right) newright = right;
+            @name@rf_recursive_@suff@(v,
+#if @arg@
+                               tosort,
+#endif
+                               newleft, newright, kth);
```

Comment:
These floating point operations scare me a little, when thinking about large arrays where `(right - left + 1) != (intp)(npy_float)(right - left + 1)`. I think this needs some careful analysis (or reading of the paper) to check that this doesn't skip elements somehow.

Reply:
Are we sure these shouldn't be integer ops?

Reply:
The point of the operation is, to quote the paper,
"Use SELECT recursively on a sample of size S to get an estimate for (K-L+1)-th smallest element into X[K], biased slightly so that (K-L+1)th element is expected to lie in the smaller set after partitioning".

So, we just want (right - left + 1) = S to be approximately equal to 0.5 * N^(2/3), which isn't an integer value, so rounding up or down doesn't make a difference.

Reply:
@hameerabbasi 
What's going on here is that we (recursively) select a subsample of size 0.5 * N^(2/3), find a pivot there, and this pivot is then approximately the right one, so using it speeds up the algorithm. Computing 0.5 * N^(2/3) is why we need floating point numbers.
npy_float Z = npy_log(N);
npy_float S = C2 * npy_exp(2*Z/3.0);

Reply:
If it’s an approximation and we’re handling the boundaries, it should be fine.

Reply:
...And this is only done for N >= 600.

Reply:
> These floating point operations scare me a little, when thinking about large arrays where `(right - left + 1) != (intp)(npy_float)(right - left + 1)`. I think this needs some careful analysis (or reading of the paper) to check that this doesn't skip elements somehow.

I'm sorry, I misunderstood you at first, I fixed the issue now:
- replaced npy_float with npy_longdouble so that big intp numbers don't get rounded off;
- when I convert back to intp, I check that the number absolute value is not greater than intp.max

I'm failing some CI doctests right now. When I try to run the doctests locally (np.test(doctests=True)), I get:
ValueError: doctests not supported. Is it supposed to be like that? Does this mean that numpy doesn't support doctests and I don't have to pass them?

Reply:
We use `python runtests.py --refguide-check` not the pytest doctests.

Reply:
> replaced npy_float with npy_longdouble so that big intp numbers don't get rounded off;

This isn't a viable solution - not all platforms have a large enough long double for this to be true.

I suspect if done correctly, these float operations _are_ safe, providing as you say they're used as an estimate, and any invariants they're intended to satisfy are enforced. This is one of those things that looks scary at a first glance, and will need a good comment in the source code to justify its correctness (once we work out whether it is correct).

So actually, using `float` (float32) is actually a good idea, because it makes it easier to check for bugs!

Reply:
Thank you! I've put back the float. Yes, these are estimates, and I check that there is no overflow when I convert float back to intp.

Reply:
@eric-wieser, @mattip, can you please review again? I returned the float. I added explanation why there is no numerical overflow in the code comments.

Reply:
Is there anything else I should do?